src/main/scala/mesosphere/marathon/core/deployment/impl/TaskReplaceActor.scala

```diff
@@ -285,20 +285,20 @@ object TaskReplaceActor extends StrictLogging {
   /** Encapsulates the logic how to get a Restart going */
   private[impl] case class RestartStrategy(nrToKillImmediately: Int, maxCapacity: Int)
 
-  private[impl] def computeRestartStrategy(runSpec: RunSpec, runningInstancesCount: Int): RestartStrategy = {
+  private[impl] def computeRestartStrategy(runSpec: RunSpec, consideredHealthyInstancesCount: Int): RestartStrategy = {
     // in addition to a spec which passed validation, we require:
     require(runSpec.instances > 0, s"instances must be > 0 but is ${runSpec.instances}")
-    require(runningInstancesCount >= 0, s"running instances count must be >=0 but is $runningInstancesCount")
+    require(consideredHealthyInstancesCount >= 0, s"running instances count must be >=0 but is $consideredHealthyInstancesCount")
```

Comment:
Hmm, with this change it's possible that all instances are unhealthy and this actor will crash now. We'll need to handle zero appropriately. @jeschkies can you weigh in?

Reply:
You're right.
From what I understand, this requirement was a safety because 0 running instances meant there was no task to replace (only task to start, which is handled by a different deployment step).

We can now have 0 healthy instance but >0 running instances.

I would naively tend to remove that assertion (following code will lead to creation of a RestartStrategy with nrToKillImmediately = 0).
Tasks from old app version will be killed later as we received healthy instance messages.

What do you think?

Reply:
Hmpf, this good is the most brittle we have in all of Marathon. @kamaradclimber, could you give an example with let's say three instances to illustrate the issue you've encountered. Cause I don't quite understood the issue your trying to fix.

Reply:
Sure!
 
Let's assume an application with 3 instances. Instances take 15minutes to pass their healthchecks.
Application configuration has a minimumHealthCapacity of 0.6 (allowing 1 instance to be killed by marathon during deployments) and maximumOverCapacity of 0 (never go beyond 3 instances).

- T+0m: a deployment of the application is initiated. We have 3 instances: A1, B1, C1
- T+0m: marathon kills one instance (A1), start a new one (A2) which is not healthy
- T+5m: marathon leader disappear, a new marathon leader is elected
- T+5m: new marathon leader kills B1, spawns B2.

At this moment, minimumHealthCapacity is violated because state is the following: A2 (not healthy), B2 (not healthy), C1 (healthy).

Later (at T+15m), A2 will pass its healthchecks and deployment will continue.

Reply:
Thanks. That helps a lot. So, it seems we do not reconcile for old instance but only already started instances. See `TaskReplaceActor:201` and `ReadinessBehaviour:195`. I actually came across this recently and was not sure which entity actually sets `instance.state.healthy`. We have Mesos and Marathon health checks. I know for sure that Mesos health checks alter the `instance.state.healthy` state. However, I don't know if Marathon health checks do the same. If they do we could simplify `ReadinessBehaviour` and the ignition strategy. Maybe that's too much for now.

Just so that we are not altering the logic to much. What do you think about moving the health filtering to the ignition strategy? Pass the seq instead of an integer.

Reply:
Thanks for your input.
I haven't thought about mesos health checks vs marathon health checks. Two pointers make me think this is ok:
- when testing this patch for real, I've used marathon health checks (so they are taken in consideration by instance.state.healthy)
- I've followed code path to confirm mesos health checks are indeed used to update `instance.state.healthy`

I'm not so sure about changing signature of `computeRestartStrategy` method, a lot of tests in RestartStrategyTest.scala file would need to be modified (they currently pass an int to the method and would need to have a list of healthy instances to be stubbed).
Any specific reason to move health filtering logic in this method?

Reply:
> Any specific reason to move health filtering logic in this method?
Yes. A lot of the other logic depends on `activeInstances`. I would rather not pull in the health checks there. If we would receive updates for health changes and have it attached to the instances it would be a different story.

Reply:
I'm not sure to see the logic depending on `activeInstances` (apart from test).

Reply:
Oh, you are right. I mixed it up with `instancesAlreadyStarted`.

Reply:
Could we list remaining actions to do on this PR, I'm a bit lost.

Reply:
> Could we list remaining actions to do on this PR, I'm a bit lost.

It's all good from my side. All you need to do is to remove the unused imports I've listed below so that the build passes.